compiler/src/typed/typecore.re

```diff
@@ -1405,6 +1433,95 @@ and type_expect_ =
       exp_type: ty_res,
       exp_env: env,
     });
+  | PExpPartial(func, part_args) =>
+    begin_def(); /* one more level for non-returning functions */
+    let funct = type_exp(env, func);
+    end_def();
+
+    let (_, _, labeled_args) =
+      process_application(
+        ~loc,
+        (part_sarg, label, _) => (label, part_sarg),
+        ~get_arg_loc=sarg => sarg.ppaa_loc,
+        ~get_arg_label=sarg => sarg.ppaa_label,
+        env,
+        funct,
+        part_args,
+      );
+    open Ast_helper;
+
+    let (app_args, new_func_params) =
+      List.split(
+        List.mapi(
+          (i, (label, arg)) => {
+            let (expr, param_pat) =
+              switch (arg.ppaa_expr) {
+              | ArgumentGiven(expr) => (expr, None)
+              | ArgumentHole(_) =>
+                let name =
+                  switch (label) {
+                  | Labeled(name)
+                  | Default(name) => name
+                  | Unlabeled => mknoloc("$arg" ++ string_of_int(i))
```

Comment:
I was having a little trouble visualizing what is actually happening here are we building a tuple from the arguments and then passing the tuple around and at the end passing the tuple values to the function when its fully applied? or did I misunderstand the abstraction here?